claim/claimstore.go

```diff
@@ -2,82 +2,427 @@ package claim
+			return Result{}, ErrResultNotFound
+		}
+		return Result{}, err
+	}
+	result := Result{}
+	err = json.Unmarshal(bytes, &result)
+	return result, err
+}
+
+func (s Store) ReadAllResults(claimID string) ([]Result, error) {
+	items, err := s.backingStore.ReadAll(ItemTypeResults, claimID)
+	if err != nil {
+		// TODO: handle claim not found
+		return nil, err
+	}
+
+	results := make(Results, len(items))
+	for i, bytes := range items {
+		var result Result
+		err = json.Unmarshal(bytes, &result)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshaling result: %v", err)
+		}
+		results[i] = result
+	}
+
+	sort.Sort(results)
+	return results, nil
+}
+
+func (s Store) ReadLastResult(claimID string) (Result, error) {
+	resultIDs, err := s.backingStore.List(ItemTypeResults, claimID)
+	if err != nil {
+		// TODO: handle installation/claim not found
+		return Result{}, err
+	}
+
+	if len(resultIDs) == 0 {
+		return Result{}, fmt.Errorf("claim %s has no results", claimID)
+	}
+
+	sort.Strings(resultIDs)
+	lastResultID := resultIDs[len(resultIDs)-1]
+
+	return s.ReadResult(lastResultID)
+}
+
+func (s Store) ReadOutput(claim Claim, result Result, outputName string) (Output, error) {
+	bytes, err := s.backingStore.Read(ItemTypeOutputs, s.outputKey(result.ID, outputName))
+	if err != nil {
+		if strings.Contains(err.Error(), crud.ErrRecordDoesNotExist.Error()) {
+			return Output{}, ErrOutputNotFound
+		}
+		return Output{}, err
+	}
+
+	sensitive, err := claim.Bundle.IsOutputSensitive(outputName)
+	if err != nil {
+		return Output{}, errors.Wrapf(err, "could not determine if the output %q is sensitive", outputName)
+	}
+
+	if sensitive {
+		bytes, err = s.decrypt(bytes)
+		if err != nil {
+			return Output{}, errors.Wrap(err, "error decrypting output")
+		}
+	}
+
+	return Output{
+		Claim:  claim,
+		Result: result,
+		Name:   outputName,
+		Value:  bytes,
+	}, nil
+}
+
+func (s Store) SaveClaim(claim Claim) error {
+	bytes, err := json.MarshalIndent(claim, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	bytes, err = s.encrypt(bytes)
+	if err != nil {
+		return errors.Wrapf(err, "error encrypting claim %s of installation %s", claim.ID, claim.Installation)
+	}
+
+	return s.backingStore.Save(ItemTypeClaims, claim.Installation, claim.ID, bytes)
+}
+
+func (s Store) SaveResult(result Result) error {
+	bytes, err := json.MarshalIndent(result, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	return s.backingStore.Save(ItemTypeResults, result.ClaimID, result.ID, bytes)
+}
+
+func (s Store) SaveOutput(output Output) error {
+	sensitive, err := output.Claim.Bundle.IsOutputSensitive(output.Name)
+	if err != nil {
+		return errors.Wrapf(err, "could not determine if the output %q is sensitive", output.Name)
+	}
+
+	data := output.Value
+	if sensitive {
+		data, err = s.encrypt(output.Value)
+		if err != nil {
+			// TODO (carolynvs) make all of the error messages provide context like this one
```

Comment:
I wanted to review all the error messages coming out of claim store and make sure if we had useful context info (like the name of the installation) that we were giving it in the error instead of giving vague errors. I'll do that and remove the TODO.

Reply:
Fixed!